File: macos/AgentCord/GrokUsage.swift
//
//  GrokUsage.swift
//  AgentCord
//
//  Polls Grok CLI / SuperGrok weekly credit usage via
//  GET https://cli-chat-proxy.grok.com/v1/billing?format=credits
//  using the OIDC tokens stored in ~/.grok/auth.json (same source as the
//  Grok CLI `/usage` view and Orca's grok-fetcher).
//
//  This is weekly included credits (`creditUsagePercent`) — NOT the per-session
//  context window fill (that still lives on GrokSession for the session card).
//

import Foundation
import Combine

final class GrokUsage: ObservableObject {

    /// Latest billing snapshot, or nil when it could not be fetched.
    @Published private(set) var current: GrokUsageInfo?

    /// True when ~/.grok/auth.json has usable OIDC credentials.
    @Published private(set) var isAuthenticated = false

    var pollInterval: TimeInterval = 300
    var minFetchInterval: TimeInterval = 60
    var maxStaleness: TimeInterval = 1800

    private var lastSuccess: Date = .distantPast
    private var lastAttempt: Date = .distantPast

    private var cachedAccessToken: String?
    private var cachedRefreshToken: String?
    private var cachedClientID: String?
    private var cachedIssuer: String?
    private var cachedUserID: String?

    private let urlSession: URLSession
    private let queue = DispatchQueue(label: "com.agentcord.grok-usage", qos: .utility)
    private var timer: DispatchSourceTimer?

    /// Must match Grok CLI / Orca: credits format returns `creditUsagePercent`.
    private static let billingURL = URL(string: "https://cli-chat-proxy.grok.com/v1/billing?format=credits")!
    private static let cliAuthHeader = "xai-grok-cli"

    private static let cacheURL: URL = {
        let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first
            ?? FileManager.default.temporaryDirectory
        let dir = base.appendingPathComponent("AgentCord", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir.appendingPathComponent("grok-usage-cache.json")
    }()

    init() {
        let config = URLSessionConfiguration.ephemeral
        config.timeoutIntervalForRequest = 15
        config.waitsForConnectivity = false
        urlSession = URLSession(configuration: config)

        if let cached = Self.loadCache(), Date().timeIntervalSince(cached.fetchedAt) <= maxStaleness {
            current = cached.info
            lastSuccess = cached.fetchedAt
        }
        isAuthenticated = Self.readAuthFile() != nil
    }

    func start() {
        let t = DispatchSource.makeTimerSource(queue: queue)
        let firstDelay: TimeInterval = (current != nil) ? 5 : 2
        t.schedule(deadline: .now() + firstDelay, repeating: pollInterval)
        t.setEventHandler { [weak self] in self?.fetch() }
        t.resume()
        timer = t
    }

    func stop() {
        timer?.cancel()
        timer = nil
    }

    func refresh() {
        queue.async { [weak self] in
            guard let self else { return }
            guard Date().timeIntervalSince(self.lastAttempt) >= self.minFetchInterval else { return }
            self.fetch()
        }
    }

    // MARK: Fetch

    private func fetch() {
        lastAttempt = Date()
        loadCredentialsFromDiskIfNeeded()

        guard cachedAccessToken != nil || cachedRefreshToken != nil else {
            publishAuth(false)
            handleFailure()
            return
        }
        publishAuth(true)
        requestBilling(allowRefresh: true)
    }

    private func requestBilling(allowRefresh: Bool) {
        guard let access = cachedAccessToken, !access.isEmpty else {
            if allowRefresh {
                refreshAccessToken { [weak self] ok in
                    guard let self else { return }
                    if ok { self.requestBilling(allowRefresh: false) }
                    else { self.handleFailure() }
                }
            } else {
                handleFailure()
            }
            return
        }

        var request = URLRequest(url: Self.billingURL)
        request.httpMethod = "GET"
        request.setValue("Bearer \(access)", forHTTPHeaderField: "Authorization")
        request.setValue("application/json", forHTTPHeaderField: "Accept")
        request.setValue(Self.cliAuthHeader, forHTTPHeaderField: "X-XAI-Token-Auth")
        request.setValue("GrokCLI", forHTTPHeaderField: "User-Agent")
        if let userID = cachedUserID, !userID.isEmpty {
            request.setValue(userID, forHTTPHeaderField: "x-userid")
        }

        urlSession.dataTask(with: request) { [weak self] data, response, _ in
            guard let self else { return }
            self.queue.async {
                let status = (response as? HTTPURLResponse)?.statusCode ?? -1
                if status == 401, allowRefresh {
                    self.refreshAccessToken { [weak self] ok in
                        guard let self else { return }
                        if ok { self.requestBilling(allowRefresh: false) }
                        else { self.handleFailure() }
                    }
                    return
                }
                guard status == 200, let data,
                      let decoded = try? JSONDecoder().decode(BillingResponse.self, from: data),
                      let info = decoded.toUsageInfo() else {
                    self.handleFailure()
                    return
                }
                self.lastSuccess = Date()
                Self.saveCache(info, fetchedAt: self.lastSuccess)
                self.publish(info)
            }
        }.resume()
    }

    private func refreshAccessToken(completion: @escaping (Bool) -> Void) {
        guard let refresh = cachedRefreshToken, !refresh.isEmpty,
              let clientID = cachedClientID, !clientID.isEmpty else {
            completion(false)
            return
        }

        let issuer = (cachedIssuer?.isEmpty == false) ? cachedIssuer! : "https://auth.x.ai"
        guard let url = URL(string: issuer.trimmingCharacters(in: CharacterSet(charactersIn: "/")) + "/oauth2/token") else {
            completion(false)
            return
        }

        var request = URLRequest(url: url)
        request.httpMethod = "POST"
        request.setValue("application/x-www-form-urlencoded", forHTTPHeaderField: "Content-Type")
        let body = [
            "grant_type": "refresh_token",
            "refresh_token": refresh,
            "client_id": clientID
        ]
        .map { key, value in
            "\(key)=\(value.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? value)"
        }
        .joined(separator: "&")
        request.httpBody = Data(body.utf8)

        urlSession.dataTask(with: request) { [weak self] data, response, _ in
            guard let self else { return }
            self.queue.async {
                guard let http = response as? HTTPURLResponse, http.statusCode == 200,
                      let data,
                      let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
                      let access = obj["access_token"] as? String, !access.isEmpty else {
                    completion(false)
                    return
                }
                self.cachedAccessToken = access
                if let newRefresh = obj["refresh_token"] as? String, !newRefresh.isEmpty {
                    self.cachedRefreshToken = newRefresh
                }
                completion(true)
            }
        }.resume()
    }

    private func handleFailure() {
        if Date().timeIntervalSince(lastSuccess) > maxStaleness {
            Self.clearCache()
            publish(nil)
        }
    }

    private func publish(_ info: GrokUsageInfo?) {
        DispatchQueue.main.async { [weak self] in
            guard let self else { return }
            if self.current != info { self.current = info }
        }
    }

    private func publishAuth(_ authenticated: Bool) {
        DispatchQueue.main.async { [weak self] in
            guard let self else { return }
            if self.isAuthenticated != authenticated { self.isAuthenticated = authenticated }
        }
    }

    // MARK: Credentials

    private func loadCredentialsFromDiskIfNeeded() {
        guard let auth = Self.readAuthFile() else {
            cachedAccessToken = nil
            cachedRefreshToken = nil
            cachedClientID = nil
            cachedIssuer = nil
            cachedUserID = nil
            return
        }
        cachedRefreshToken = auth.refreshToken
        cachedClientID = auth.clientID
        cachedIssuer = auth.issuer
        cachedUserID = auth.userID
        if let access = auth.accessToken, !access.isEmpty {
            cachedAccessToken = access
        }
    }

    private struct AuthTokens {
        var accessToken: String?
        var refreshToken: String?
        var clientID: String?
        var issuer: String?
        var userID: String?
    }

    private static func authFileURL() -> URL? {
        if let env = ProcessInfo.processInfo.environment["GROK_HOME"], !env.isEmpty {
            let url = URL(fileURLWithPath: env, isDirectory: true)
                .appendingPathComponent("auth.json")
            if FileManager.default.fileExists(atPath: url.path) { return url }
        }
        let home = FileManager.default.homeDirectoryForCurrentUser
            .appendingPathComponent(".grok/auth.json")
        return FileManager.default.fileExists(atPath: home.path) ? home : nil
    }

    /// auth.json is a map of account-key → credential object.
    private static func readAuthFile() -> AuthTokens? {
        guard let url = authFileURL(),
              let data = try? Data(contentsOf: url),
              let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any]
        else { return nil }

        // Prefer the first object that looks like an OIDC credential entry.
        for (_, value) in root {
            guard let obj = value as? [String: Any] else { continue }
            let access = obj["key"] as? String
            let refresh = obj["refresh_token"] as? String
            if (access == nil || access?.isEmpty == true),
               (refresh == nil || refresh?.isEmpty == true) {
                continue
            }
            return AuthTokens(
                accessToken: access,
                refreshToken: refresh,
                clientID: obj["oidc_client_id"] as? String,
                issuer: obj["oidc_issuer"] as? String,
                userID: obj["user_id"] as? String
            )
        }
        return nil
    }

    // MARK: Disk cache

    private struct CachePayload: Codable {
        var fetchedAt: Date
        var info: GrokUsageInfo
    }

    private static func loadCache() -> CachePayload? {
        guard let data = try? Data(contentsOf: cacheURL) else { return nil }
        return try? JSONDecoder().decode(CachePayload.self, from: data)
    }

    private static func saveCache(_ info: GrokUsageInfo, fetchedAt: Date) {
        let payload = CachePayload(fetchedAt: fetchedAt, info: info)
        guard let data = try? JSONEncoder().encode(payload) else { return }
        try? data.write(to: cacheURL, options: .atomic)
    }

    private static func clearCache() {
        try? FileManager.default.removeItem(at: cacheURL)
    }
}

// MARK: - Models

/// Grok weekly subscription credit usage (CLI / SuperGrok billing).
struct GrokUsageInfo: Equatable, Codable {
    /// Weekly included credits window (`creditUsagePercent`).
    var weekly: UsageInfo.Window
    /// On-demand window when a cap is set.
    var onDemand: UsageInfo.Window?
    var billingPeriodStart: Date?
    var billingPeriodEnd: Date?
}

// MARK: - Wire format

private struct BillingResponse: Decodable {
    struct MoneyVal: Decodable {
        let val: Double?
    }

    struct Period: Decodable {
        let type: String?
        let start: String?
        let end: String?
    }

    struct Config: Decodable {
        let creditUsagePercent: Double?
        let currentPeriod: Period?
        let onDemandCap: MoneyVal?
        let onDemandUsed: MoneyVal?
        let billingPeriodStart: String?
        let billingPeriodEnd: String?
    }

    let config: Config?
    /// Some responses flatten credit fields at the top level.
    let creditUsagePercent: Double?

    func toUsageInfo() -> GrokUsageInfo? {
        let cfg = config
        let percentRaw = cfg?.creditUsagePercent ?? creditUsagePercent
        guard let percentRaw, percentRaw.isFinite else { return nil }

        let percent = min(100, max(0, Int(percentRaw.rounded())))
        let periodEnd = Self.parseISO(cfg?.currentPeriod?.end ?? cfg?.billingPeriodEnd)
        let periodStart = Self.parseISO(cfg?.currentPeriod?.start ?? cfg?.billingPeriodStart)

        let weekly = UsageInfo.Window(
            percent: percent,
            severity: Self.severity(for: percent),
            resetsAt: periodEnd
        )

        var onDemand: UsageInfo.Window?
        if let cap = cfg?.onDemandCap?.val, cap > 0 {
            let odUsed = cfg?.onDemandUsed?.val ?? 0
            let odPercent = min(100, max(0, Int((odUsed / cap * 100).rounded())))
            onDemand = UsageInfo.Window(
                percent: odPercent,
                severity: Self.severity(for: odPercent),
                resetsAt: periodEnd
            )
        }

        return GrokUsageInfo(
            weekly: weekly,
            onDemand: onDemand,
            billingPeriodStart: periodStart,
            billingPeriodEnd: periodEnd
        )
    }

    private static func severity(for percent: Int) -> String {
        if percent >= 95 { return "critical" }
        if percent >= 80 { return "warning" }
        return "normal"
    }

    private static let isoFractional: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        return f
    }()

    private static let isoPlain: ISO8601DateFormatter = {
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime]
        return f
    }()

    private static func parseISO(_ string: String?) -> Date? {
        guard let string, !string.isEmpty else { return nil }
        return isoFractional.date(from: string) ?? isoPlain.date(from: string)
    }
}

File: windows/GrokUsage.cs
// Polls Grok CLI / SuperGrok weekly credit usage via
// GET https://cli-chat-proxy.grok.com/v1/billing?format=credits
// using the OIDC tokens stored in ~/.grok/auth.json (same source as the
// Grok CLI `/usage` view).
//
// This is weekly included credits (`creditUsagePercent`) — NOT the per-session
// context window fill (that still lives on GrokSession for the session card).
// Port of AgentCord/GrokUsage.swift.

using System.Globalization;
using System.IO;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text.Json;

namespace AgentCord;

public sealed class GrokUsage : IDisposable
{
    public GrokUsageInfo? Current { get; private set; }

    /// <summary>True when ~/.grok/auth.json has usable OIDC credentials.</summary>
    public bool IsAuthenticated { get; private set; }

    /// <summary>Email of the signed-in xAI account, recorded in ~/.grok/auth.json.</summary>
    public string? AccountEmail { get; private set; }

    public TimeSpan PollInterval { get; init; } = TimeSpan.FromSeconds(300);
    public TimeSpan MinFetchInterval { get; init; } = TimeSpan.FromSeconds(60);
    /// <summary>Keep a disk-cached snapshot for a day so relaunch / idle
    /// stretches still show the last known numbers.</summary>
    public TimeSpan MaxStaleness { get; init; } = TimeSpan.FromHours(24);

    private static readonly Uri BillingUrl = new("https://cli-chat-proxy.grok.com/v1/billing?format=credits");
    private const string CliAuthHeader = "xai-grok-cli";

    private readonly HttpClient _http = new() { Timeout = TimeSpan.FromSeconds(15) };
    private readonly string? _grokHomeOverride;
    private readonly object _lock = new();
    private DateTime _lastSuccess = DateTime.MinValue;
    private DateTime _lastAttempt = DateTime.MinValue;
    private string? _cachedAccessToken;
    private string? _cachedRefreshToken;
    private string? _cachedClientId;
    private string? _cachedIssuer;
    private string? _cachedUserId;
    private System.Threading.Timer? _timer;

    public GrokUsage(string? grokHome = null)
    {
        _grokHomeOverride = grokHome;
        if (LoadCache() is { } cached
            && DateTime.UtcNow - cached.FetchedAt <= MaxStaleness)
        {
            Current = cached.Info;
            _lastSuccess = cached.FetchedAt;
        }
        var auth = ReadAuthFile();
        IsAuthenticated = auth is not null;
        AccountEmail = auth?.Email;
    }

    public void Start() => SetEnabled(true);

    public void SetEnabled(bool enabled)
    {
        if (enabled)
        {
            if (_timer is not null) return;
            var first = Current is null ? TimeSpan.FromSeconds(2) : TimeSpan.FromSeconds(5);
            _timer = new System.Threading.Timer(_ => _ = FetchAsync(), null, first, PollInterval);
            return;
        }
        _timer?.Dispose();
        _timer = null;
    }

    public void Refresh()
    {
        lock (_lock)
        {
            if (DateTime.UtcNow - _lastAttempt < MinFetchInterval) return;
        }
        _ = FetchAsync();
    }

    public void Dispose()
    {
        _timer?.Dispose();
        _http.Dispose();
    }

    private async Task FetchAsync()
    {
        lock (_lock) _lastAttempt = DateTime.UtcNow;

        try
        {
            LoadCredentialsFromDisk();
            if (_cachedAccessToken is null && _cachedRefreshToken is null)
            {
                IsAuthenticated = false;
                AccountEmail = null;
                HandleFailure();
                return;
            }

            IsAuthenticated = true;
            var info = await RequestBillingAsync(allowRefresh: true);
            if (info is null)
            {
                HandleFailure();
                return;
            }

            lock (_lock) _lastSuccess = DateTime.UtcNow;
            Current = info;
            SaveCache(info, _lastSuccess);
        }
        catch
        {
            HandleFailure();
        }
    }

    private async Task<GrokUsageInfo?> RequestBillingAsync(bool allowRefresh)
    {
        if (string.IsNullOrEmpty(_cachedAccessToken))
        {
            if (allowRefresh && await RefreshAccessTokenAsync())
                return await RequestBillingAsync(allowRefresh: false);
            return null;
        }

        using var request = new HttpRequestMessage(HttpMethod.Get, BillingUrl);
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", _cachedAccessToken);
        request.Headers.Accept.Add(new MediaTypeWithQualityHeaderValue("application/json"));
        request.Headers.TryAddWithoutValidation("X-XAI-Token-Auth", CliAuthHeader);
        request.Headers.TryAddWithoutValidation("User-Agent", "GrokCLI");
        if (!string.IsNullOrEmpty(_cachedUserId))
            request.Headers.TryAddWithoutValidation("x-userid", _cachedUserId);

        using var response = await _http.SendAsync(request);
        if ((int)response.StatusCode == 401 && allowRefresh && await RefreshAccessTokenAsync())
            return await RequestBillingAsync(allowRefresh: false);
        if (!response.IsSuccessStatusCode) return null;

        var body = await response.Content.ReadAsStringAsync();
        return ParseBilling(body);
    }

    private async Task<bool> RefreshAccessTokenAsync()
    {
        if (string.IsNullOrEmpty(_cachedRefreshToken) || string.IsNullOrEmpty(_cachedClientId))
            return false;

        var issuer = string.IsNullOrEmpty(_cachedIssuer) ? "https://auth.x.ai" : _cachedIssuer.TrimEnd('/');
        var url = issuer + "/oauth2/token";

        using var request = new HttpRequestMessage(HttpMethod.Post, url)
        {
            Content = new FormUrlEncodedContent(new Dictionary<string, string>
            {
                ["grant_type"] = "refresh_token",
                ["refresh_token"] = _cachedRefreshToken,
                ["client_id"] = _cachedClientId,
            }),
        };

        using var response = await _http.SendAsync(request);
        if (!response.IsSuccessStatusCode) return false;

        try
        {
            using var doc = JsonDocument.Parse(await response.Content.ReadAsStringAsync());
            var access = doc.RootElement.TryGetProperty("access_token", out var accessEl)
                && accessEl.ValueKind == JsonValueKind.String
                ? accessEl.GetString()
                : null;
            if (string.IsNullOrEmpty(access)) return false;
            _cachedAccessToken = access;
            if (doc.RootElement.TryGetProperty("refresh_token", out var refreshEl)
                && refreshEl.ValueKind == JsonValueKind.String
                && refreshEl.GetString() is { Length: > 0 } refresh)
            {
                _cachedRefreshToken = refresh;
            }
            return true;
        }
        catch
        {
            return false;
        }
    }

    private void HandleFailure()
    {
        lock (_lock)
        {
            if (DateTime.UtcNow - _lastSuccess > MaxStaleness)
            {
                Current = null;
                ClearCache();
            }
        }
    }

    private void LoadCredentialsFromDisk()
    {
        var auth = ReadAuthFile();
        if (auth is null)
        {
            _cachedAccessToken = null;
            _cachedRefreshToken = null;
            _cachedClientId = null;
            _cachedIssuer = null;
            _cachedUserId = null;
            AccountEmail = null;
            return;
        }

        _cachedRefreshToken = auth.RefreshToken;
        _cachedClientId = auth.ClientId;
        _cachedIssuer = auth.Issuer;
        _cachedUserId = auth.UserId;
        AccountEmail = string.IsNullOrEmpty(auth.Email) ? null : auth.Email;
        if (!string.IsNullOrEmpty(auth.AccessToken))
            _cachedAccessToken = auth.AccessToken;
    }

    private sealed record AuthTokens(
        string? AccessToken,
        string? RefreshToken,
        string? ClientId,
        string? Issuer,
        string? UserId,
        string? Email);

    private AuthTokens? ReadAuthFile()
    {
        try
        {
            var path = AuthFilePath();
            if (path is null) return null;
            using var doc = JsonDocument.Parse(File.ReadAllText(path));
            if (doc.RootElement.ValueKind != JsonValueKind.Object) return null;

            foreach (var prop in doc.RootElement.EnumerateObject())
            {
                if (prop.Value.ValueKind != JsonValueKind.Object) continue;
                var obj = prop.Value;
                var access = StringProp(obj, "key");
                var refresh = StringProp(obj, "refresh_token");
                if (string.IsNullOrEmpty(access) && string.IsNullOrEmpty(refresh)) continue;
                return new AuthTokens(
                    access,
                    refresh,
                    StringProp(obj, "oidc_client_id"),
                    StringProp(obj, "oidc_issuer"),
                    StringProp(obj, "user_id"),
                    StringProp(obj, "email"));
            }
        }
        catch
        {
            // Missing or malformed auth is treated as signed out.
        }
        return null;
    }

    private string? AuthFilePath()
    {
        if (!string.IsNullOrEmpty(_grokHomeOverride))
        {
            var overridePath = Path.Combine(_grokHomeOverride, "auth.json");
            return File.Exists(overridePath) ? overridePath : null;
        }
        if (Environment.GetEnvironmentVariable("GROK_HOME") is { Length: > 0 } env)
        {
            var envPath = Path.Combine(env, "auth.json");
            if (File.Exists(envPath)) return envPath;
        }
        var home = Path.Combine(
            Environment.GetFolderPath(Environment.SpecialFolder.UserProfile),
            ".grok", "auth.json");
        return File.Exists(home) ? home : null;
    }

    /// <summary>Parse a billing JSON body into weekly / on-demand windows.</summary>
    public static GrokUsageInfo? ParseBilling(string json)
    {
        try
        {
            using var doc = JsonDocument.Parse(json);
            return ParseBilling(doc.RootElement);
        }
        catch
        {
            return null;
        }
    }

    internal static GrokUsageInfo? ParseBilling(JsonElement root)
    {
        JsonElement? config = root.TryGetProperty("config", out var cfg) && cfg.ValueKind == JsonValueKind.Object
            ? cfg
            : null;

        string? periodEndRaw = null;
        if (config is JsonElement c)
        {
            if (c.TryGetProperty("currentPeriod", out var period) && period.ValueKind == JsonValueKind.Object)
                periodEndRaw = StringProp(period, "end");
            periodEndRaw ??= StringProp(c, "billingPeriodEnd");
        }
        periodEndRaw ??= StringProp(root, "billingPeriodEnd");

        var periodEndMs = ParseIsoMs(periodEndRaw);

        // `creditUsagePercent` is present with the real figure on SuperGrok
        // weekly accounts. Unified-billing accounts omit it entirely but still
        // report the weekly `currentPeriod` — the Grok CLI shows that as 0%
        // used, so treat an absent percent (when we have a period) as zero.
        var percentRaw = NumberProp(config, "creditUsagePercent")
            ?? NumberProp(root, "creditUsagePercent")
            ?? (periodEndMs is not null ? 0 : null);
        if (percentRaw is not double percentValue || double.IsNaN(percentValue) || double.IsInfinity(percentValue))
            return null;

        var percent = ClampPercent(percentValue);
        var weekly = MakeWindow(percent, periodEndMs);

        UsageWindow? onDemand = null;
        if (config is JsonElement cfgObj
            && MoneyVal(cfgObj, "onDemandCap") is double cap && cap > 0)
        {
            var used = MoneyVal(cfgObj, "onDemandUsed") ?? 0;
            onDemand = MakeWindow(ClampPercent(used / cap * 100.0), periodEndMs);
        }

        return new GrokUsageInfo { Weekly = weekly, OnDemand = onDemand };
    }

    private static UsageWindow MakeWindow(int percent, long? resetsAtMs) => new()
    {
        Percent = percent,
        Severity = percent >= 95 ? "critical" : percent >= 80 ? "warning" : "normal",
        ResetsAtMs = resetsAtMs,
    };

    private static int ClampPercent(double value) =>
        Math.Min(100, Math.Max(0, (int)Math.Round(value)));

    private static double? NumberProp(JsonElement? obj, string name)
    {
        if (obj is not JsonElement el) return NumberProp(default(JsonElement), name);
        return NumberProp(el, name);
    }

    private static double? NumberProp(JsonElement obj, string name)
    {
        if (obj.ValueKind != JsonValueKind.Object) return null;
        if (!obj.TryGetProperty(name, out var v) || v.ValueKind != JsonValueKind.Number)
            return null;
        return v.GetDouble();
    }

    private static double? MoneyVal(JsonElement obj, string name)
    {
        if (!obj.TryGetProperty(name, out var v)) return null;
        if (v.ValueKind == JsonValueKind.Number) return v.GetDouble();
        if (v.ValueKind == JsonValueKind.Object)
            return NumberProp(v, "val");
        return null;
    }

    private static string? StringProp(JsonElement obj, string name) =>
        obj.TryGetProperty(name, out var value) && value.ValueKind == JsonValueKind.String
            ? value.GetString()
            : null;

    private static long? ParseIsoMs(string? value)
    {
        if (string.IsNullOrEmpty(value)) return null;
        if (DateTimeOffset.TryParse(
                value, CultureInfo.InvariantCulture,
                DateTimeStyles.AssumeUniversal | DateTimeStyles.AdjustToUniversal,
                out var dto))
        {
            return dto.ToUnixTimeMilliseconds();
        }
        return null;
    }

    // --- Disk cache

    private sealed record CachePayload(DateTime FetchedAt, GrokUsageInfo Info);

    private static string CachePath
    {
        get
        {
            var baseDir = Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData);
            if (string.IsNullOrEmpty(baseDir)) baseDir = Path.GetTempPath();
            return Path.Combine(baseDir, "AgentCord", "grok-usage-cache.json");
        }
    }

    private static CachePayload? LoadCache()
    {
        try
        {
            return JsonSerializer.Deserialize<CachePayload>(File.ReadAllText(CachePath));
        }
        catch
        {
            return null;
        }
    }

    private static void SaveCache(GrokUsageInfo info, DateTime fetchedAt)
    {
        try
        {
            var path = CachePath;
            Directory.CreateDirectory(Path.GetDirectoryName(path)!);
            File.WriteAllText(path, JsonSerializer.Serialize(new CachePayload(fetchedAt, info)));
        }
        catch
        {
            // Best-effort cache.
        }
    }

    private static void ClearCache()
    {
        try { File.Delete(CachePath); }
        catch { }
    }
}
